proposals/0000-dependent-type-design.rst

```diff
@@ -0,0 +1,388 @@
+
+The history of these proposals suggest that we indeed have been worried about criterion
+2 for some time, without ever being very explicit about it. This current proposal
+is about making this choice more explicit -- and committing to continue to honor
+criterion 2 going forward.
+
+Motivation for dependent types
+------------------------------
+Dependent types would allow Haskellers to encode more invariants in their
+types, allow more flexible (often heterogeneous) data structures, and allow
+for the possibility of more code optimizations. Given the availability of
+the `singletons`_ library, which simulates dependent types and has 91 reverse
+dependencies, many of these
+examples are possible in Haskell today. However, dependent types are far
+from easy to use today, and the overarching goal of the proposals that would
+be affected by this current one is to make them easier to work with.
+
+* Chapter 3 of `my thesis`_ is all about motivating dependent types in Haskell.
+* `Why Dependent Types Matter <http://www.cs.nott.ac.uk/~psztxa/publ/ydtm.pdf>`_
+* `The Power of Pi <https://cs.ru.nl/~wouters/Publications/ThePowerOfPi.pdf>`_
+* `Constrained Type Families <https://richarde.dev/papers/2017/partiality/partiality.pdf>`_ and `Partial Type Constructors <https://richarde.dev/papers/2020/partialdata/partialdata.pdf>`_ would fit better in a language with dependent types; the latter explicitly desugars into a dependently typed language.
+* `Stitch <https://richarde.dev/papers/2018/stitch/stitch.pdf>`_ uses techniques from dependent types to implement a lambda-calculus interpreter that is well-typed by construction.
+* `Dependent Types in Haskell <https://www.youtube.com/watch?v=J8iitZtNeJk>`_, a talk by
+  Stephanie Weirich on how to encode well-formed regular expressions with dependent types.
+* `A Reflection on Types <https://richarde.dev/papers/2016/dynamic/dynamic.pdf>`_, on dynamic typing in Haskell, relying on dependent-type machinery. Expansions of this idea
+  will require even more power in the type system.
+* Though I do not have an easily-separable example, the use of dependent types
+  allow us to drop tags in certain scenarios: if the type invariants indicate
+  that only one disjunct of a union type is possible, then we can skip the runtime
+  check for that type.
+* The `singletons paper
+  <https://richarde.dev/papers/2012/singletons/paper.pdf>`_ contains an
+  example of well-typed database access using dependent types; it would be
+  possible to skip certain dynamic type checks if we could rely on the
+  dependent types instead.
+
+Any reader is invited to add more links to this list via a pull request.
+
+Proposed Change Specification
+-----------------------------
+The GHC committee would consider compatibility with ergonomic dependent types
```

Comment:
I think this proposal needs to be a bit more specific about what "ergonomic dependent types" would look like in Haskell, if we're going to use that as a criterion when evaluating other proposals. Is your thesis still an up-to-date source for the overall design you have in mind?

Reply:
Mostly, yes -- with a very notable exception that my thesis doesn't tackle the namespace problem. It uses `'` to promote names from the term level to the type level. I no longer think we should do that.

There is a bullet under Examples attempting to define aspects of "ergonomic dependent types". I, too, worry that this is too ill-defined for a proposal such as this, but I'm hoping my key example -- don't syntactically distinguish between terms and types -- is helpful.